Allow templates to refer to other side of multi-part cards

haltman-at opened the pull request:
Closes #171.  **Note:  This PR alters the card update script, so after merging this you'll have to rerun that script.**

This PR adds a new field to rule objects, the `side` field.  This field can take one of three values.  If it is left `undefined` (or is absent), then the rule applies to the card as normal.  (Originally this was going to be called `side: "normal"`, but that would have caused compatibility issues with existing presets.)  If it is set to `"other-side"`, then this rule will instead be checked against the card's other side -- if there is no other side, the rule automatically fails to match.  Finally, if it is set to `"other-front-meld"`... well, I'll detail that behavior, but it's something special that can be used for meld cards; if the card is not a meld card, the rule will automatically fail to match.

Templates that use other-side rules (or other-front-meld rules) cannot be saved in presets.  This is basically for the same reason that presets can't be placed in or-groups, namely, the current system for card matching, based on preset expansion rather than recursive evaluation, wouldn't be able to handle it.  Well, that's not entirely true -- what it can't handle is having a preset that uses an other-side rule, which then has the other-side modifier *applied to it in a template*.

This is for two reasons.  The first, as mentioned, is that this doesn't fit well with the current system of preset expansion rather than recursive evaluation.  Still, it might be doable with some effort, but then there's the second issue: Composability.  For this to work, we'd have to figure out what it means to compose two "other-side" modifiers.  Well, OK, the other-side of the other-side is the original side, surely?  Yes, but then you throw in other-front-meld and everything becomes an inconsistent mess.  So, this is a no-go.

As such, it makes sense to either bar the use of other-side rules in presets, or the application of other-side modifiers to presets.  Since there's a lot more reason to do the latter than the former, I've chosen the former to prohibit.  Attempting to save such a rule in a preset will result in an error before the preset naming box even pops up.  (There's also a server-side error for if you circumvent this check.)  While I was at it, I realized that validation errors should be checked at this stage too -- no reason to wait until you've named it! -- so I moved that check earlier as well.  Doing this meant creating the actual template earlier, so to communicate it to the template naming box submit method, I store it in a global variable in the meantime.  Maybe not the best way, but it works.

Anyway, time to answer the question: How does all this work with meld cards?  For other two-sided cards the application is straightforward, but meld cards are necessarily a little wonky.  I think the final result is pretty reasonable, though.

The way it works is that if you use "other-side" rules, then for a meld back, all other-side rules must be satisfied by the *same* meld front; however, it can be *either* of the fronts.  So, for instance, searching for a meld back whose other side is a Phyrexian Human will turn up Mishra, Lost to Phyrexia, as will searching for a meld back whose other side is a Phyrexian Dragon; but searching for a meld back whose other side is a Human Dragon will turn up nothing.

What about "other-front-meld"?  For a meld front, this refers to its counterpart that it melds with.  For a meld *back*, it refers to either front side, but it always refers to whichever front is *not* covered by other-side.  That is to say, for a meld back, if you have both other-side rules and other-front-meld rules, these two side labels will alwasy refer to the two different fronts, but which one is which can vary.  Thus, for instance, searching for a meld back whose other-side is a Human and whose other-front-meld is a Dragon *will* turn up Mishra, Lost to Phyrexia; and if you swap the roles, making the other-side a Dragon, and the other-front-meld a Human, the search will still work and return the same results -- for a meld back, the roles have no intrinsic meaning, their only meaning is relative to one another.

This leaves the interface.  I'll admit, it's not great, but, uh, maybe you have ideas for how to improve it.  (Part of the awkwardness is my lack of knowledge of CSS, sorry.)  Anyway, each rule now has, next to the remove button, an "other side button".  By default, rules have `side: undefined` (obviously).  Clicking the "other side" button will swap between it being a normal rule and an other-side rule.  To get an other-front-meld rule, you must *shift*-click this button.  Sorry, that's not exactly great UI, but I figured that other-front-meld is going to be pretty rare and not something you want coming up normally... I dunno, suggestions for a better interface is very much appreciated.  From other-front-meld, a normal click will return it to normal, while a shift-click will return it to other-side.

(For the button's icon, I used a 🔄 emoji.  Well, not the actual emoji, but rather an image of it, which seemed to work better.  That should be fine to use, right?)

Next to the button is some text that displays its state, showing whether the rule applies to "This card", "Other side", or "Add'l front" (maybe I didn't need to abbreviate that).  Perhaps a better graphic designer could have incorporated this information into the icon itself with several different icons, but I'm not a graphic designer and that seemed hard, so text it is.  (Btw, when creating OR groups, one may click this text just as one may click the rule's label.)

Now, I have to warn you that there is some **serious hackery** associated with this text!  This is probably the worst thing in this PR.  You see, above I described things in terms of a `side` field.  Thing is... the template isn't actually created until `createTemplate` is called!  Which means that when it's called.  All those `side` fields have to be populated somehow.  Where are they populated from?  **That's right: they're populated from the displayed text.**  This is obviously hacky, but I didn't really want to move the `createTemplate` call earlier, so there you go.  I realized later an alternative, less hacky solution would have been to use a hidden input element or something, but, oh well.  If you want me to go back and make that change, I can do so.

(Speaking of hacks, note that adding these two elements to each rule required adjusting some absolute indices... yikes!  And if the hidden input field were added, I'd have to adjust them again.  But that's existing hackery.)

Finally, one more bit of hackery: In order to make rules read grammatically, I made it so that the message includes "'s" on the end if the rule is not a preset, this involves more hackery, but I'll skip describing that here.  For presets this is absent, and the result is somewhat awkward due to capitalization, but oh well.  Lowercasing things would introduce a new can of worms (you're going to have problems if preset descriptions are not displayed exactly as entered!), so, I left it at this, despite the awkwardness.

Finally, one last hack worth mentioning: In some cases, for reasons I don't understand, setting `textContext` didn't work, so I set `innerHTML` instead.  Don't worry, it's not being set to any sort of user-entered value!  But hm, wonder what's up with that.

### Testing

I tested this with every type of multipart card.  In addition, I tested that this works when applied to a preset, that it works in an or group, that it errors when saved in a preset (though I didn't test the server-side check), and that all the meld stuff works as described above.  I also tested that saving new presets still works after my alterations to it here.

KingSupernova31 commented:
> Originally this was going to be called side: "normal", but that would have caused compatibility issues with existing presets.

Note this isn't a big deal in general, if you ever need to change the format of something in the db, just give me a function that accepts a question object and outputs the reformatted one, and I'll run it on the production database as part of the update.

> Anyway, each rule now has, next to the remove button, an "other side button". By default, rules have side: undefined (obviously). Clicking the "other side" button will swap between it being a normal rule and an other-side rule. To get an other-front-meld rule, you must shift-click this button.

Why not make it a dropdown menu? I think that would look nicer and be more intuitive, along with being consistent with all the other fields.

> Where are they populated from? That's right: they're populated from the displayed text. This is obviously hacky, but I didn't really want to move the createTemplate call earlier, so there you go.

This doesn't actually seem all that hacky to me; where else is the information supposed to come from? You could store it in a separate variable, but then that's redundant with the variable that's actually being displayed in the DOM. Using a normal <select> element rather than a custom system would also fix this though.

> (Speaking of hacks, note that adding these two elements to each rule required adjusting some absolute indices... yikes! And if the hidden input field were added, I'd have to adjust them again. But that's existing hackery.)

Huh, I forgot about that. No idea why I did that in such a dumb way rather than just assigning them each a class name. I'll fix that later.

KingSupernova31 commented:
> Finally, one last hack worth mentioning: In some cases, for reasons I don't understand, setting textContext didn't work, so I set innerHTML instead. Don't worry, it's not being set to any sort of user-entered value! But hm, wonder what's up with that.

I tested it on my device and `textContent` seems to work properly, in both Chrome and Firefox. Is it possible that you have an extension in the browser you're testing it on that interferes?

haltman-at commented:
> Note this isn't a big deal in general, if you ever need to change the format of something in the db, just give me a function that accepts a question object and outputs the reformatted one, and I'll run it on the production database as part of the update.

Oh, OK.  I think I'll just leave this one as-is, but good to know we can do that.  Of course here it's the presets rather than the questions that would need updating.  (Note btw that there are some other spots in the code where I relied on the fact that `undefined` is the default, those would need updating, but I've marked those.)
 
> Why not make it a dropdown menu? I think that would look nicer and be more intuitive, along with being consistent with all the other fields.

Oh, that's a way better idea.  We discussed the button idea earlier and I kind of fixated on that, I guess.  I'll go change that.

> This doesn't actually seem all that hacky to me; where else is the information supposed to come from? You could store it in a separate variable, but then that's redundant with the variable that's actually being displayed in the DOM. Using a normal `<select>` element rather than a custom system would also fix this though.

Sure does!

haltman-at commented:
> > Finally, one last hack worth mentioning: In some cases, for reasons I don't understand, setting textContext didn't work, so I set innerHTML instead. Don't worry, it's not being set to any sort of user-entered value! But hm, wonder what's up with that.
> 
> I tested it on my device and `textContent` seems to work properly, in both Chrome and Firefox. Is it possible that you have an extension in the browser you're testing it on that interferes?

Huh, interesting question.  I'd be surprised, but it could be the case.  I'm on Firefox 141.0.2 on Linux Mint.  The only two extensions I have active are HTTPS Everywhere and User-Agent Switcher and Manager.  I guess I could try disabling those and see if there's any effect.

KingSupernova31 commented:
> Of course here it's the presets rather than the questions that would need updating.

Oh right, derp. Well anyway same idea holds, changing production data files is something I'm happy to do, just make it easy for me.

haltman-at commented:
OK, updated it to use more straightforward drop-down menus instead of the silly button system.  I've also changed it to use `"normal"` instead of `undefined`.  (This includes updating `defaultPresetTemplates.json`, but not the data files; those will need to be updated via a script.)  Requesting rereview.

I did some quick retesting to make sure things still basically worked, but didn't do a thorough check like before.

Those `innerHTML`s that I added are now gone, so I'll investigate those some other time if it comes up again.

Here's the quick script I used to update the data files: (you'll notice that it was also used to update `defaultPresetTemplates.json`, but since that's in source control and included in the PR, that bit's been commented out)

```javascript
const fs = require("fs");

const updateJSON = presets => presets.map(
	preset => ({ ...preset, rules: preset.rules.map(rule => ({ ...rule, side: "normal" }))})
);

/*
const defaults = JSON.parse(fs.readFileSync("./defaultPresetTemplates.json", "utf8"));
const updatedDefaults = updateJSON(defaults);
const updatedDefaultsJson = JSON.stringify(updatedDefaults, undefined, "\t");
fs.writeFileSync("./defaultPresetTemplates.json", updatedDefaultsJson);
*/

const presets = JSON.parse(fs.readFileSync("./data_files/presetTemplates.json", "utf8"));
const updatedPresets = updateJSON(presets);
const updatedPresetsJson = JSON.stringify(updatedPresets);
fs.writeFileSync("./data_files/presetTemplates.json", updatedPresetsJson);
fs.writeFileSync("./public_html/public_data_files/presetTemplates.js", preparePresetLoader(updatedPresetsJson));

function preparePresetLoader(presetsJson) {
	return `
const presetTemplates = ${presetsJson};

let presetIds = [];
let presetDescriptions = [];
for (let preset of presetTemplates) {
	presetIds.push(preset.id);
	presetDescriptions.push(preset.description);
}
if (presetIds.length !== Array.from(new Set(presetIds)).length) {
	handleError("Duplicate preset IDs");
}
if (presetDescriptions.length !== Array.from(new Set(presetDescriptions)).length) {
	handleError("Duplicate preset descriptions");
}

if (typeof module === "object") {
	module.exports = presetTemplates;
};
`
}
```

KingSupernova31 commented:
You forgot to change the index reference for the expanding rules text input field, now it expands the operator dropdown instead. You can ignore that though, I'll fix it when I switch to class names.

The logic doesn't seem to work as described? Rules text contains flying + rules text contains vigilance is generating brisela, despite neither front side having flying and vigilance.f

Also could you change "add'l front" to "other front"? Just looks cleaner I think.

haltman-at commented:
(Fixed the bug, requesting rereview.  Also changed "add'l front" to "second front".)
